src/iter/walk_tree.rs

```diff
@@ -0,0 +1,533 @@
+        let right = right_children
+            .map(|mut c| {
+                std::mem::swap(&mut c, &mut self.to_explore);
+                WalkTreePrefixProducer {
+                    to_explore: c,
+                    seen: Vec::new(),
+                    breed: self.breed,
+                    phantom: PhantomData,
+                }
+            })
+            .or_else(|| {
+                // we can still try to divide 'seen'
+                let right_seen = split_vec(&mut self.seen);
+                right_seen.map(|s| WalkTreePrefixProducer {
+                    to_explore: Default::default(),
+                    seen: s,
+                    breed: self.breed,
+                    phantom: PhantomData,
+                })
+            });
+        (self, right)
+    }
+    fn fold_with<F>(mut self, mut folder: F) -> F
+    where
+        F: Folder<Self::Item>,
+    {
+        // start by consuming everything seen
+        for s in self.seen {
+            folder = folder.consume(s);
+            if folder.full() {
+                return folder;
+            }
+        }
+        // now do all remaining explorations
+        while let Some(e) = self.to_explore.pop() {
+            self.to_explore.extend((self.breed)(&e));
+            folder = folder.consume(e);
+            if folder.full() {
+                return folder;
+            }
+        }
+        folder
+    }
+}
+
+/// ParallelIterator for arbitrary tree-shaped patterns.
+/// Returned by the [`walk_tree_prefix()`] function.
+///
+/// [`walk_tree_prefix()`]: fn.walk_tree_prefix.html
+#[derive(Debug)]
+pub struct WalkTreePrefix<S, B, I> {
+    initial_state: S,
+    breed: B,
+    phantom: PhantomData<I>,
+}
+
+impl<S, B, I> ParallelIterator for WalkTreePrefix<S, B, I>
+where
+    S: Send,
+    B: Fn(&S) -> I + Send + Sync,
+    I: IntoIterator<Item = S> + Send,
+{
+    type Item = S;
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+    where
+        C: UnindexedConsumer<Self::Item>,
+    {
+        let producer = WalkTreePrefixProducer {
+            to_explore: once(self.initial_state).collect(),
+            seen: Vec::new(),
+            breed: &self.breed,
+            phantom: PhantomData,
+        };
+        bridge_unindexed(producer, consumer)
+    }
+}
+
+/// Create a tree-like prefix parallel iterator from an initial root node.
+/// The `children_of` function should take a node and return an iterator over its child nodes.
+/// The best parallelization is obtained when the tree is balanced
+/// but we should also be able to handle harder cases.
+///
+/// # Ordering
+///
+/// This function guarantees a prefix ordering. See also [`walk_tree_postfix`],
+/// which guarantees a postfix order.
+/// If you don't care about ordering, you should use [`walk_tree`],
+/// which will use whatever is believed to be fastest.
+/// Between siblings, children are reduced in reverse order -- that is, the children that returned last are reduced first.
```

Comment:
The reverse order seems really bizarre -- is that necessary? As a user, I would expect a "normal" pre-order traversal of a binary tree to first visit the node, then the left children, then the right children.

Reply:
hi, this is the fastest version. since it is a prefix order we need to consume the father before its children. but we need to borrow the father to iterate on children. this means we first need to extend the stack with the children.
because it's a vec the last pushed will be the first poped and this reverses children's order.

other options are:
- use a more complex struct and some unsafe.
- require the iterator to be double ended
- first collect the iterator into a vec and then extend the stack in the right order

to me the only acceptable option is the third one. it comes with a performance hit though.

if you think it would be better to have the standard order then i can re-implement this version.

Reply:
ok, so i did a quick bench (the one in rayon_demo)
this is the current version:
sum: 382ns collect: 683ns
this is the version with the "fixed" ordering:
sum: 705ns collect: 950ns

Reply:
That speed difference seems pretty significant to me

Reply:
> first collect the iterator into a vec and then extend the stack in the right order

You could do this in-place by noting the previous length, extending, then reversing the new tail slice. That's still some added cost, but at least it doesn't require a new allocation.

Reply:
I do care about the speed, but that comes after having useful semantics. Reversed child order seems unusual enough to me that I doubt anyone really *wants* that order, and if they don't care they can use the unordered `walk_tree`.

Reply:
in my bench i actually did it without an extra allocation. using a buffer vec and draining / extending.
i'm not sure that the actual order is that bad. user can just reverse their iterator if they really care about siblings ordering.
walk_tree has no guarantee while this has

Reply:
i did it like this :
```
         self.reorder_buffer.extend((self.breed)(&e));
         self.to_explore.extend(self.reorder_buffer.drain(..).rev());
```

Reply:
just to be clear, i do understand there is a choice to be made here.
the overhead is between 10 and 15 nanoseconds per element. this is not zero cost but not excessive.
if you ask me i would favor the current (reversed) order because the algorithm is "natural" this way. user can also reverse their iterators to get to the other order. however i do see why the other option also makes sense.

in the end, choose whatever you want and let me know.

Reply:
Ideally, internal algorithm details should not be so visible in the public API, especially since those details might want to change in the future. I want to aim for what is most "natural" to the user first, without exposing details like "well actually you probably want to reverse this." That's even evident in your `tree_prefix_collect` benchmark where you have an explanatory comment, "large indices to the left, small to the right", being different than the others.

To that end, I think `DoubleEndedIterator` might be reasonable, although you immediately dismissed that. The user can provide the children in their natural order, **matching the visitation order**, but we'll have the flexibility to internally iterate in reverse. Are there realistic use-cases that you think could not be double-ended?

Of course there are plenty of non-DE iterators, but I'm asking for a real data structure that could not operate this way, especially since you say "users can also reverse their iterators." If they can so easily reverse, then DE should be no issue; if they can't, then I think we won't be serving them well with an algorithmically-reversed order.

Reply:
so i did the switch to double ended iterators

Reply:
I think we now just need to update (or remove) this line about reversal.